File: test/modules/linear_boltzmann_solvers/transport_transient/transient_transport_1d_3.lua
-- 1D Transient Transport test with Vacuum BC.
-- SDM: PWLD
-- Test:
num_procs = 2

--############################################### Check num_procs
if check_num_procs == nil and number_of_processes ~= num_procs then
  log.Log(
    LOG_0ERROR,
    "Incorrect amount of processors. "
      .. "Expected "
      .. tostring(num_procs)
      .. ". Pass check_num_procs=false to override if possible."
  )
  os.exit(false)
end

--############################################### Setup mesh
nodes = {}
N = 2000
L = 100.0
xmin = -L / 2
dx = L / N
for i = 1, (N + 1) do
  k = i - 1
  nodes[i] = xmin + k * dx
end

meshgen1 = mesh.OrthogonalMeshGenerator.Create({ node_sets = { nodes } })
mesh.MeshGenerator.Execute(meshgen1)

--############################################### Set Material IDs
mesh.SetUniformMaterialID(0)

vol0 = logvol.RPPLogicalVolume.Create({ infx = true, infy = true, zmin = -L / 4, zmax = L / 4 })
mesh.SetMaterialIDFromLogicalVolume(vol0, 1)

--############################################### Add materials
materials = {}
materials[1] = mat.AddMaterial("Strong fuel")
materials[2] = mat.AddMaterial("Weak fuel")

-- Define microscopic cross sections
xs_strong_fuel_micro = xs.Create()
xs.Set(xs_strong_fuel_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_k1_6_1g.xs")
xs_weak_fuelA_micro = xs.Create()
xs.Set(xs_weak_fuelA_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_critical_1g.xs")
xs_weak_fuelB_micro = xs.Create()
xs.Set(xs_weak_fuelB_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_weak_1g.xs")

atom_density = 0.056559
xs_strong_fuel = xs.MakeScaled(xs_strong_fuel_micro, atom_density) --critical
xs_weak_fuelA = xs.MakeScaled(xs_weak_fuelA_micro, atom_density) --critical
xs_weak_fuelB = xs.MakeScaled(xs_weak_fuelB_micro, atom_density) --critical

num_groups = 1
mat.SetProperty(materials[1], TRANSPORT_XSECTIONS, EXISTING, xs_strong_fuel)
mat.SetProperty(materials[2], TRANSPORT_XSECTIONS, EXISTING, xs_weak_fuelA)

mat.SetProperty(materials[1], ISOTROPIC_MG_SOURCE, FROM_ARRAY, { 0.0 })
mat.SetProperty(materials[2], ISOTROPIC_MG_SOURCE, FROM_ARRAY, { 0.0 })

function SwapXS(solver_handle, new_xs)
  mat.SetProperty(materials[2], TRANSPORT_XSECTIONS, EXISTING, new_xs)
  lbs.InitializeMaterials(solver_handle)
end

--############################################### Setup Physics
phys1 = LBSCreateTransientSolver()

--========== Groups
grp = {}
for g = 1, num_groups do
  grp[g] = LBSCreateGroup(phys1)
end

--========== ProdQuad
pquad = aquad.CreateProductQuadrature(GAUSS_LEGENDRE, 16)

--========== Groupset def
gs0 = LBSCreateGroupset(phys1)
cur_gs = gs0
LBSGroupsetAddGroups(phys1, cur_gs, 0, num_groups - 1)
LBSGroupsetSetQuadrature(phys1, cur_gs, pquad)
LBSGroupsetSetAngleAggDiv(phys1, cur_gs, 1)
LBSGroupsetSetGroupSubsets(phys1, cur_gs, 8)
LBSGroupsetSetIterativeMethod(phys1, cur_gs, KRYLOV_GMRES)
--LBSGroupsetSetIterativeMethod(phys1,cur_gs,KRYLOV_RICHARDSON)
LBSGroupsetSetResidualTolerance(phys1, cur_gs, 1.0e-6)
LBSGroupsetSetMaxIterations(phys1, cur_gs, 1000)
LBSGroupsetSetGMRESRestartIntvl(phys1, cur_gs, 100)
--LBSGroupsetSetWGDSA(phys1,cur_gs,30,1.0e-4,false," ")
--LBSGroupsetSetTGDSA(phys1,cur_gs,30,1.0e-4,false," ")

--
----############################################### Set boundary conditions
--bsrc={}
--for g=1,num_groups do
--    bsrc[g] = 0.0
--end
--bsrc[1] = 1.0/2
--LBSSetProperty(phys1,BOUNDARY_CONDITION,ZMIN,LBSBoundaryTypes.REFLECTING);
--LBSSetProperty(phys1,BOUNDARY_CONDITION,ZMAX,LBSBoundaryTypes.REFLECTING);
--
LBSSetProperty(phys1, DISCRETIZATION_METHOD, PWLD)
LBSSetProperty(phys1, SCATTERING_ORDER, 1)

LBKESSetProperty(phys1, "MAX_ITERATIONS", 1000)
LBKESSetProperty(phys1, "TOLERANCE", 1.0e-8)

LBSSetProperty(phys1, USE_PRECURSORS, true)

--LBSSetProperty(phys1, VERBOSE_INNER_ITERATIONS, false)
LBSSetProperty(phys1, VERBOSE_INNER_ITERATIONS, false)
LBSSetProperty(phys1, VERBOSE_OUTER_ITERATIONS, true)

--############################################### Initialize and Execute Solver
solver.Initialize(phys1)

LBTSSetProperty(phys1, "TIMESTEP", 1e-3 * 100)
LBTSSetProperty(phys1, "TIMESTOP", 1.0 * 100)
LBTSSetProperty(phys1, "MAX_TIMESTEPS", -1)
LBTSSetProperty(phys1, "VERBOSITY_LEVEL", 0)
LBTSSetProperty(phys1, "TIMESTEP_METHOD", "CRANK_NICHOLSON")

phys1name = solver.GetName(phys1)
initial_FR = lbs.ComputeFissionRate(phys1, "OLD")

--time = 0.0
--psi_t = psi_0
--for k=1,10 do
--    solver.Step(phys1)
--
--    FRf = lbs.ComputeFissionRate(phys1,"NEW") --time+dt
--    FRi = lbs.ComputeFissionRate(phys1,"OLD") --time
--    dt = LBTSGetProperty(phys1, "TIMESTEP")
--    time = LBTSGetProperty(phys1, "TIME")
--    new_time = time+dt
--
--    period = dt/math.log(FRf/FRi)
--    log.Log(LOG_0, string.format("%s %4d time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
--            phys1name,k,time,dt,period,FRf/initial_FR))
--
--    if (not timestep_rejected) then
--        LBTSAdvanceTimeValues()
--    end
--    LBTSAdvanceTimeData(phys1)
--end

--time = 0.0
--time_stop = 1.0*100
--k=0
--swapped = false
--timestep_rejected = false
--
--tolA = 10.0
--while (time < time_stop) do
--    solver.Step(phys1)
--    FRf = lbs.ComputeFissionRate(phys1,"NEW")
--    FRi = lbs.ComputeFissionRate(phys1,"OLD")
--    dt = LBTSGetProperty(phys1, "TIMESTEP")
--    time = LBTSGetProperty(phys1, "TIME")
--    period = dt/math.log(FRf/FRi)
--
--    if (time >= 0.2 and not swapped) then
--        SwapXS(phys1, xs_weak_fuelB)
--        swapped = true
--    end
--
--    if (not timestep_rejected) then
--        LBTSAdvanceTimeData(phys1)
--        k = k + 1
--        log.Log(LOG_0, string.format("%s %4d time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
--                phys1name,k,time,dt,period,FRf/initial_FR))
--    else
--        timestep_rejected = false
--    end
--end
swapped = false
function MyCallBack()
  FRf = lbs.ComputeFissionRate(phys1, "NEW")
  FRi = lbs.ComputeFissionRate(phys1, "OLD")
  dt = LBTSGetProperty(phys1, "TIMESTEP")
  time = LBTSGetProperty(phys1, "TIME")
  period = dt / math.log(FRf / FRi)

  if time >= 0.2 and not swapped then
    SwapXS(phys1, xs_weak_fuelB)
    swapped = true
  end
  log.Log(
    LOG_0,
    string.format(
      "%s time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
      phys1name,
      time,
      dt,
      period,
      FRf / initial_FR
    )
  )
end
LBTSSetProperty(phys1, "CALLBACK", "MyCallBack")
solver.Execute(phys1)

File: test/python/modules/linear_boltzmann_solvers/transport_transient/transient_transport_1d_3.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# 1D Transient Transport test with Vacuum BC.
# SDM: PWLD
# Test:

import os
import sys
import math

if "opensn_console" not in globals():
    from mpi4py import MPI
    size = MPI.COMM_WORLD.size
    rank = MPI.COMM_WORLD.rank
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../../../")))
    from pyopensn.mesh import OrthogonalMeshGenerator, KBAGraphPartitioner
    from pyopensn.xs import MultiGroupXS
    from pyopensn.source import VolumetricSource
    from pyopensn.aquad import GLProductQuadrature1DSlab
    from pyopensn.solver import DiscreteOrdinatesSolver, SteadyStateSolver
    from pyopensn.fieldfunc import FieldFunctionGridBased
    from pyopensn.fieldfunc import FieldFunctionInterpolationLine, FieldFunctionInterpolationVolume
    from pyopensn.settings import EnableCaliper
    from pyopensn.math import Vector3
    from pyopensn.logvol import RPPLogicalVolume

if __name__ == "__main__":


    num_procs = 2

    if size != num_procs:
        sys.exit(f"Incorrect number of processors. Expected {num_procs} processors but got {size}.")

    # Setup mesh
    nodes = []
    N = 2000
    L = 100.0
    xmin = -L / 2
    dx = L / N
    for i in range(N+1):
      nodes.append(xmin + i * dx)

    meshgen = OrthogonalMeshGenerator(node_sets = [nodes])
    grid = meshgen.Execute()

    # Set block IDs
    grid.SetUniformBlockID(0)

    vol0 = RPPLogicalVolume( infx = True, infy = True, zmin = -L / 4, zmax = L / 4 )
    grid.SetBlockIDFromLogicalVolume(vol0, 1, True)

    # Add materials
    materials = []
    materials[1] = mat.AddMaterial("Strong fuel")
    materials[2] = mat.AddMaterial("Weak fuel")

    # Define microscopic cross sections
    xs_strong_fuel_micro = xs.Create()
    xs.Set(xs_strong_fuel_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_k1_6_1g.xs")
    xs_weak_fuelA_micro = xs.Create()
    xs.Set(xs_weak_fuelA_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_critical_1g.xs")
    xs_weak_fuelB_micro = xs.Create()
    xs.Set(xs_weak_fuelB_micro, OPENSN_XSFILE, "tests/transport_transient/xs_inf_weak_1g.xs")

    atom_density = 0.056559
    xs_strong_fuel = xs.MakeScaled(xs_strong_fuel_micro, atom_density) #critical
    xs_weak_fuelA = xs.MakeScaled(xs_weak_fuelA_micro, atom_density) #critical
    xs_weak_fuelB = xs.MakeScaled(xs_weak_fuelB_micro, atom_density) #critical

    num_groups = 1
    mat.SetProperty(materials[1], TRANSPORT_XSECTIONS, EXISTING, xs_strong_fuel)
    mat.SetProperty(materials[2], TRANSPORT_XSECTIONS, EXISTING, xs_weak_fuelA)

    mat.SetProperty(materials[1], ISOTROPIC_MG_SOURCE, FROM_ARRAY, { 0.0 )
    mat.SetProperty(materials[2], ISOTROPIC_MG_SOURCE, FROM_ARRAY, { 0.0 )

    function SwapXS(solver_handle, new_xs)
      mat.SetProperty(materials[2], TRANSPORT_XSECTIONS, EXISTING, new_xs)
      InitializeMaterials(solver_handle)

    # Setup Physics
    phys = LBSCreateTransientSolver()

    #========== Groups
    grp = []
    for g in range(1, num_groups+1):
      grp[g] = LBSCreateGroup(phys)

    #========== ProdQuad
    pquad = ProductQuadrature(GAUSS_LEGENDRE, 16)

    #========== Groupset def
    gs0 = LBSCreateGroupset(phys)
    cur_gs = gs0
    LBSGroupsetAddGroups(phys, cur_gs, 0, num_groups - 1)
    LBSGroupsetSetQuadrature(phys, cur_gs, pquad)
    LBSGroupsetSetAngleAggDiv(phys, cur_gs, 1)
    LBSGroupsetSetGroupSubsets(phys, cur_gs, 8)
    LBSGroupsetSetIterativeMethod(phys, cur_gs, KRYLOV_GMRES)
    #LBSGroupsetSetIterativeMethod(phys,cur_gs,KRYLOV_RICHARDSON)
    LBSGroupsetSetResidualTolerance(phys, cur_gs, 1.0e-6)
    LBSGroupsetSetMaxIterations(phys, cur_gs, 1000)
    LBSGroupsetSetGMRESRestartIntvl(phys, cur_gs, 100)
    #LBSGroupsetSetWGDSA(phys,cur_gs,30,1.0e-4,False," ")
    #LBSGroupsetSetTGDSA(phys,cur_gs,30,1.0e-4,False," ")

    #
    #-- Set boundary conditions
    #bsrc=[]
    #for g in range(1, num_groups+1):
    #    bsrc[g] = 0.0
    #end
    #bsrc[1] = 1.0/2
    #LBSSetProperty(phys,BOUNDARY_CONDITION,ZMIN,LBSBoundaryTypes.REFLECTING);
    #LBSSetProperty(phys,BOUNDARY_CONDITION,ZMAX,LBSBoundaryTypes.REFLECTING);
    #
    LBSSetProperty(phys, DISCRETIZATION_METHOD, PWLD)
    LBSSetProperty(phys, SCATTERING_ORDER, 1)

    LBKESSetProperty(phys, "MAX_ITERATIONS", 1000)
    LBKESSetProperty(phys, "TOLERANCE", 1.0e-8)

    LBSSetProperty(phys, USE_PRECURSORS, True)

    #LBSSetProperty(phys, VERBOSE_INNER_ITERATIONS, False)
    LBSSetProperty(phys, VERBOSE_INNER_ITERATIONS, False)
    LBSSetProperty(phys, VERBOSE_OUTER_ITERATIONS, True)

    # Initialize and Execute Solver
    solver.Initialize(phys)

    LBTSSetProperty(phys, "TIMESTEP", 1e-3 * 100)
    LBTSSetProperty(phys, "TIMESTOP", 1.0 * 100)
    LBTSSetProperty(phys, "MAX_TIMESTEPS", -1)
    LBTSSetProperty(phys, "VERBOSITY_LEVEL", 0)
    LBTSSetProperty(phys, "TIMESTEP_METHOD", "CRANK_NICHOLSON")

    physname = solver.GetName(phys)
    initial_FR = ComputeFissionRate(phys, "OLD")

    #time = 0.0
    #psi_t = psi_0
    #for k in range(1, 10+1):
    #    solver.Step(phys)
    #
    #    FRf = ComputeFissionRate(phys,"NEW") --time+dt
    #    FRi = ComputeFissionRate(phys,"OLD") --time
    #    dt = LBTSGetProperty(phys, "TIMESTEP")
    #    time = LBTSGetProperty(phys, "TIME")
    #    new_time = time+dt
    #
    #    period = dt/math.log(FRf/FRi)
    #    log.Log(LOG_0, string.format("%s %4d time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
    #            physname,k,time,dt,period,FRf/initial_FR))
    #
    #    if (not timestep_rejected) then
    #        LBTSAdvanceTimeValues()
    #    end
    #    LBTSAdvanceTimeData(phys)
    #end

    #time = 0.0
    #time_stop = 1.0*100
    #k=0
    #swapped = False
    #timestep_rejected = False
    #
    #tolA = 10.0
    #while (time < time_stop) do
    #    solver.Step(phys)
    #    FRf = ComputeFissionRate(phys,"NEW")
    #    FRi = ComputeFissionRate(phys,"OLD")
    #    dt = LBTSGetProperty(phys, "TIMESTEP")
    #    time = LBTSGetProperty(phys, "TIME")
    #    period = dt/math.log(FRf/FRi)
    #
    #    if (time >= 0.2 and not swapped) then
    #        SwapXS(phys, xs_weak_fuelB)
    #        swapped = True
    #    end
    #
    #    if (not timestep_rejected) then
    #        LBTSAdvanceTimeData(phys)
    #        k = k + 1
    #        log.Log(LOG_0, string.format("%s %4d time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
    #                physname,k,time,dt,period,FRf/initial_FR))
    #    else
    #        timestep_rejected = False
    #    end
    #end
    swapped = False
    function MyCallBack()
      FRf = ComputeFissionRate(phys, "NEW")
      FRi = ComputeFissionRate(phys, "OLD")
      dt = LBTSGetProperty(phys, "TIMESTEP")
      time = LBTSGetProperty(phys, "TIME")
      period = dt / math.log(FRf / FRi)

      if time >= 0.2 and not swapped then
        SwapXS(phys, xs_weak_fuelB)
        swapped = True
      log.Log(
        LOG_0,
        string.format(
          "%s time=%10.3g dt=%10.4g period=%10.3g FR=%10.3e",
          physname,
          time,
          dt,
          period,
          FRf / initial_FR
        )
      )

    LBTSSetProperty(phys, "CALLBACK", "MyCallBack")
    solver.Execute(phys)
